### PR TITLE
gpmovemirrors: remove tablespace directories from old host

### PR DESCRIPTION
**Pull request by dpandhi-git**
The issue was raised on the internal portal for utility gpmovemirrors that failed to remove old user-defined tablespace directories for a mirror when it's moved to a new host. 

This commit includes the following changes:

  * common function `get_tablespace_locations` to get user-defined tablespace locations from all hosts or specific mirror data directory

  * unit and behave test for function 

  * Function is implemented in utilities gpmovemirrors and gpdeletesystem to fix the issue

[Pipeline](https://cm.ci.gpdb.pivotal.io/teams/main/pipelines/gpmovemirror_rm_old_tblspc_dir)

Co-authored-by: Jamie McAtamney jmcatamney@vmware.com, Orhan Kislal okislal@vmware.com

### CODE REVIEW

**Comment by kalensk** [comment]
Awesome!

1) We should consider having some behave tests? 

2) I wonder if there is already a python function or something that does `getsegmentTablespaceDirs`? If not, it might be useful to put it in a common location for other utilities to leverage?

File: gpMgmt/bin/gppylib/operations/getSegmentTablespaceDir.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+                                ON t.gp_segment_id = c.content
+                                AND c.role='m' AND c.datadir ='{data_directory}'
+                            """ .format(oid_subq=oid_subq, data_directory=data_directory)
+
+                    res = dbconn.query(conn, TABLESPACE_LOCATION)
+                    for r in res:
+                        self.tablespace_dir.append(r)
+
+        except:
+            logger.info("Unable to fetch tablespace directory")
```
[the remark this answers is not in the document]
Comment (dpandhi-git) [reply gpMgmt/bin/gppylib/operations/getSegmentTablespaceDir.py]: As discussed removed exception block to make it consistent with other code. As the only exception expected was for DB unable to connect

File: gpMgmt/bin/gppylib/operations/test/unit/test_unit_getSegmentTablespaceDir.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+import unittest
+from mock import MagicMock, Mock, mock_open, patch
+from gppylib.operations.getSegmentTablespaceDir import GetTablespaces
+
+class GetTablespaceDirTestCase(unittest.TestCase):
+
+    def setUp(self):
+        self.tablespace = GetTablespaces()
+
+    @patch('gppylib.operations.getSegmentTablespaceDir.dbconn.query')
```
[the remark this answers is not in the document]
Comment (dpandhi-git) [reply gpMgmt/bin/gppylib/operations/test/unit/test_unit_getSegmentTablespaceDir.py]: updated unit tests to have mock in setUp to avoid the repetition

**Review by kaknikhil — APPROVED** [review]

File: gpMgmt/bin/gpmovemirrors [hunk 3]
```diff
@@ -440,11 +446,23 @@ try:
...
     if numberOfWorkers > 0:
         pool = WorkerPool(numWorkers=numberOfWorkers)
         for mirror in mirrorsToDelete:
-            logger.info("About to remove old mirror segment directory: " + mirror.dataDirectory)
+            logger.info("About to remove old mirror segment directories: " + mirror.dataDirectory)
+
+            if mirror_tablespace_locations[mirror.dataDirectory]:
+                for host, content, tablespace_dir in mirror_tablespace_locations[mirror.dataDirectory]:
+                    if (host, content) not in mirrorDirectories:
+                        cmd = RemoveDirectory("remove old mirror tablespace directory", tablespace_dir, ctxt=REMOTE,
```
[the remark this answers is not in the document]
Comment (dpandhi-git) [reply gpMgmt/bin/gpmovemirrors]: Done

**Review by kaknikhil — APPROVED** [review]

**Pull request merged**